contracts/xasset-orchestrator/src/orchestrator.rs

```diff
@@ -0,0 +1,165 @@
+    fn init(&mut self, xlm_sac: Address, xlm_contract: Address) -> Result<(), Error> {
+        Contract::require_auth();
+        self.xlm_sac.set(&xlm_sac);
+        self.xlm_contract.set(&xlm_contract);
+        Ok(())
+    }
+
+    #[allow(clippy::too_many_arguments)]
+    fn deploy_asset_contract(
+        &mut self,
+        asset_contract: Address,
+        pegged_asset: Symbol,
+        min_collat_ratio: u32,
+        name: String,
+        symbol: String,
+        decimals: u32,
+        annual_interest_rate: u32,
+    ) -> Result<Address, Error> {
+        let env = env();
+        Contract::require_auth();
+        // Check if the asset contract is already deployed
+        if self.assets.has(symbol.clone()) {
+            return Err(Error::AssetAlreadyDeployed);
+        }
+
+        // Upload the xasset wasm to prep for deploy
+        let wasm_hash = env.deployer().upload_contract_wasm(xasset::WASM);
+
+        // Deploy the contract, salting with the symbol
+        let deployed_contract = create_contract(env, &wasm_hash, symbol.clone());
+
+        let contract_address = deployed_contract.map_err(|_| Error::InitFailed)?;
+        // Create a client instance for this contract so we can initialize it
+        let client = xasset::Client::new(env, &contract_address);
+
+        // Initialize the contract
+        client.cdp_init(
+            &self.xlm_sac.get().unwrap(),
+            &self.xlm_contract.get().unwrap(),
+            &asset_contract,
+            &pegged_asset,
+            &min_collat_ratio,
+            &name,
+            &symbol,
+            &decimals,
+            &annual_interest_rate,
+        );
+
+        // Store the deployed contract address in the assets map
+        self.assets.set(symbol.clone(), &contract_address);
+
+        Ok(contract_address)
+    }
+
+    fn get_asset_contract(&self, asset_symbol: String) -> Result<Address, Error> {
+        if !self.assets.has(asset_symbol.clone()) {
+            return Err(Error::NoSuchAsset);
+        }
+        Ok(self.assets.get(asset_symbol.clone()).unwrap())
+    }
+
+    fn set_asset_contract(
+        &mut self,
+        asset_symbol: String,
+        asset_contract: Address,
+    ) -> Result<(), Error> {
+        Contract::require_auth();
+        if self.assets.has(asset_symbol.clone()) {
+            return Err(Error::AssetAlreadyDeployed);
+        }
+        self.assets.set(asset_symbol.clone(), &asset_contract);
+        Ok(())
+    }
+
+    fn set_existing_asset_contract(
+        &mut self,
+        asset_symbol: String,
+        asset_contract: Address,
+    ) -> Result<(), Error> {
+        Contract::require_auth();
+        self.assets.set(asset_symbol.clone(), &asset_contract);
+        Ok(())
+    }
+}
+
+pub fn create_contract(
+    e: &Env,
+    token_wasm_hash: &BytesN<32>,
+    asset_symbol: String,
+) -> Result<Address, Error> {
+    let mut salt = Bytes::new(e);
+    salt.append(&asset_symbol.to_xdr(e));
```

Comment:
You can pass by reference preventing the need for the clone in the caller.
Normally this would be a `&str` in rust, but this is a special String type for soroban.
```suggestion
    asset_symbol: &String,
) -> Result<Address, Error> {
    let mut salt = Bytes::new(e);
    salt.append(asset_symbol.to_xdr(e));
```

Reply:
Trying to do this (`&string` argument change) causes weirdness when trying to call `.to_xdr` "expected &Bytes, found Bytes" so I'm going to revert this change

Reply:
Trying to borrow leads to ```cannot move out of `*asset_symbol` which is behind a shared reference
move occurs because `*asset_symbol` has type `loam_sdk::soroban_sdk::String`, which does not implement the `Copy` traitrustc[Click for full compiler diagnostic](rust-analyzer-diagnostics-view:/diagnostic%20message%20[0]?0#file:///Users/pselle/Dev/equitx/equitx-project/contracts/xasset-orchestrator/src/orchestrator.rs)```